From 1d9393307f4f414bd9ab7e43b39fddaf9d08a6ce Mon Sep 17 00:00:00 2001
From: Zengruan Ye <yezengruan@huawei.com>
Date: Thu, 15 Jul 2021 11:33:46 +0800
Subject: [PATCH] arm64: cpufeature: TWED support detection
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 1d9393307f4f414bd9ab7e43b39fddaf9d08a6ce
Modified-by-SEL: Yes, refreshed due to context change

virt inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I40FGG
CVE: NA

-----------------------------

TWE Delay is an optional feature in ARMv8.6 Extensions. This patch
detect this feature.

Signed-off-by: Zengruan Ye <yezengruan@huawei.com>
Signed-off-by: Jingyi Wang <wangjingyi11@huawei.com>
Reviewed-by: Keqian Zhu <zhukeqian1@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
Signed-off-by: Chenxi Mao <chenxi.mao@suse.com>
---
 arch/arm64/Kconfig               |   10 ++++++++++
 arch/arm64/include/asm/cpucaps.h |    1 +
 arch/arm64/kernel/cpufeature.c   |   12 ++++++++++++
 3 files changed, 23 insertions(+)

--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -1824,6 +1824,16 @@ config ARM64_EPAN
 	 if the cpu does not implement the feature.
 endmenu
 
+menu "ARMv8.6 architectural features"
+
+config ARM64_TWED
+	bool "Enable support for delayed trapping of WFE"
+	default y
+	help
+	  Delayed Trapping of WFE (part of the ARMv8.6 Extensions)
+
+endmenu
+
 config ARM64_SVE
 	bool "ARM Scalable Vector Extension support"
 	default y
--- a/arch/arm64/include/asm/cpucaps.h
+++ b/arch/arm64/include/asm/cpucaps.h
@@ -68,6 +68,7 @@
 #define ARM64_WORKAROUND_2457168		60
 #define ARM64_WORKAROUND_HISI_HIP08_RU_PREFETCH	61
 #define ARM64_CLEARPAGE_STNP			62
+#define ARM64_HAS_TWED				63
 
 #define ARM64_HAS_ECV				65
 #define ARM64_HAS_EPAN				66
--- a/arch/arm64/kernel/cpufeature.c
+++ b/arch/arm64/kernel/cpufeature.c
@@ -2221,6 +2221,18 @@ static const struct arm64_cpu_capabiliti
 		.type = ARM64_CPUCAP_SYSTEM_FEATURE,
 		.matches = can_clearpage_use_stnp,
 	},
+#ifdef CONFIG_ARM64_TWED
+	{
+		.desc = "Delayed Trapping of WFE",
+		.capability = ARM64_HAS_TWED,
+		.type = ARM64_CPUCAP_SYSTEM_FEATURE,
+		.matches = has_cpuid_feature,
+		.sys_reg = SYS_ID_AA64MMFR1_EL1,
+		.field_pos = ID_AA64MMFR1_TWED_SHIFT,
+		.sign = FTR_UNSIGNED,
+		.min_field_value = 1,
+	},
+#endif
 	{},
 };
 
